lib/pymedphys/_experimental/streamlit/apps/sum_doses.py

```diff
@@ -0,0 +1,120 @@
+"""Copyright (C) 2021 Matthew Jennings
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import copy
+import pathlib
+
+from pymedphys._imports import streamlit as st
+
+import numpy as np
+
+import pydicom
+from pydicom.errors import InvalidDicomError
+
+from pymedphys._dicom.coords import xyz_axes_from_dataset
+from pymedphys._dicom.dose import dose_from_dataset
+from pymedphys._streamlit import categories
+
+STREAMLIT_STATIC_PATH = pathlib.Path(st.__path__[0]) / "static"
+DOWNLOADS_PATH = STREAMLIT_STATIC_PATH / "downloads"
+if not DOWNLOADS_PATH.is_dir():
+    DOWNLOADS_PATH.mkdir()
+SUMMED_DOSE_PATH = DOWNLOADS_PATH / "RD.Summed.dcm"
+
+CATEGORY = categories.PRE_ALPHA
+TITLE = "Sum Coincident DICOM Doses"
+
+HERE = pathlib.Path(__file__).parent.resolve()
+
+
+def _axes_in_datasets_match(datasets):
+
+    # Quick shape (sanity) check
+    if not all(
+        ds.pixel_array.shape == datasets[0].pixel_array.shape for ds in datasets
+    ):
+        return False
+
+    # Full coord check:
+    all_concat_axes = [np.concatenate(xyz_axes_from_dataset(ds)) for ds in datasets]
+
+    return all(np.allclose(a, all_concat_axes[0]) for a in all_concat_axes)
+
+
+def _check_files_valid(files):
+    datasets = []
+
+    for fh in files:
+        try:
+            ds = pydicom.dcmread(fh)
+        except InvalidDicomError:
+            st.error(f"{fh.name} is not a valid DICOM file")
+            return None
+
+        if not ds.Modality == "RTDOSE":
+            st.error(f"File {fh.name} is not a valid DICOM RT Dose file")
+            return None
+
+        datasets.append(ds)
+
+    return datasets
+
+
+def sum_doses_in_datasets(datasets):
+
+    ds_summed = copy.deepcopy(datasets[0])
+
+    ds_summed.DoseSummationType = "MULTI_PLAN"
```

Comment:
I haven't seen this sort of thing before. I imagine this area of things @sjswerdloff might have some nice suggestions.

Reply:
the choice of "MULTI_PLAN" for DoseSummationType is only valid if each of the input doses have DoseSummationType of PLAN.

If you have other DoseSummationTypes on input, you may need to scale them based on number of fractions specified in the relevant fraction group in the plan (if they are of type XXX_SESSION (BEAM, FRACTION, BRACHY).

For a first pass, I would probably limit input to type PLAN (which implies dose for the whole course of treatment, not just a single fraction).
It depends on your intended use.
Its one thing to want to be able to sum something basic like a primary treatment and a boost, it's a little more complicated when you are summing from different modalities (brachy plus photon, photon and proton boost, proton and carbon boost), but it's a whole different story if you are trying to do weighted optimisation of control points or beams (e.g. re-weighting a VMAT based on ~360 "beams" and varying your optimisation criteria).

Reply:
you at least need to confirm that the doses are not from different patients.
PatientID is probably good enough (all have same PatientID).

Reply:
Both are now done :)

In the case of PLAN dose, I've included a check in `sum_doses_in_datasets()` (which will hopefully move to the `dicom` library module) and `_check_files_valid()` (which is specific to this app). It's nice to include the latter because it'll notify the user of the first file(name) that fails the requirement.